src/engine/renderer/gl_shader.h

```diff
@@ -74,20 +74,26 @@ class GLHeader
 
 	std::string                    _name;
 	std::string                    _text;
+	uint32_t                       _lineCount;
 	GLShaderManager               *_shaderManager;
 
 public:
-	GLHeader() : _name(), _text(), _shaderManager(nullptr)
+	GLHeader() : _name(), _text(), _lineCount(), _shaderManager( nullptr )
 	{}
 
 	GLHeader( const std::string &name, const std::string &text, GLShaderManager *manager ) :
 		_name( name ),
-		_text(text),
+		_text( text ),
 		_shaderManager( manager )
-	{}
+	{
+		size_t lineCount = std::count( text.begin(), text.end(), '\n' );
```

Comment:
They either added a newline if empty or affected which line reported by the driver, I'll add a comment for whichever it was.

Reply:
So this is actually because each `\n` creates 2 lines so any header with at least 1 needs to have its line count increased by 1.

Reply:
The driver can't know whether the input had an empty string concatenated in it. Maybe it's just the "compat header" adding one by being empty, canceling out the -1 subtracted from all the line count sums.

Reply:
So I checked everything again and it turns that what was throwing me off were the shader compile macros. They're never passed to `InitShader()`, but they are added when the shader code is actually sent to the driver to compile. This seems to have been a problem for a while actually, resulting in shaders with non-empty permutations not being cached. Since this behaviour is the same on master w. r. t. #insert, I've removed the erroneus line calculation, but the bug resulting from macros is still there.

Reply:
I've opened an issue for it: #1339. I consider that out of scope here though, since this pr is about logging.